src/gba/hardware-extensions.c

```diff
@@ -0,0 +1,339 @@
+                }
+            }
+        } else {
+            if (value & HWEX_CNT_FLAG_CALL) { // check call the extension
+                *returnCode = handlers->onCall(gba);
+                if (*returnCode == HWEX_RET_OK || *returnCode >= HWEX_RET_ERR_UNKNOWN) { // execution finished
+                    *cnt = 0;
+                } else { // processing
+                    *cnt = currentValue | HWEX_CNT_FLAG_CALL | HWEX_CNT_FLAG_PROCESSING;
+                }
+            }
+        }
+    }
+}
+
+void GBAHardwareExtensionsIOWrite(struct GBA* gba, uint32_t address, uint16_t value) {
+    switch (address) {
+        case REG_HWEX_ENABLE:
+            gba->hwExtensions.enabled = value == 0xC0DE;
+            break;
+        // Enable flags
+        case REG_HWEX_ENABLE_FLAGS_0:
+        case REG_HWEX_ENABLE_FLAGS_1:
+        case REG_HWEX_ENABLE_FLAGS_2:
+        case REG_HWEX_ENABLE_FLAGS_3:
+        case REG_HWEX_ENABLE_FLAGS_4:
+        case REG_HWEX_ENABLE_FLAGS_5: {
+            uint32_t index = (address - REG_HWEX_ENABLE_FLAGS_0) >> 1;
+            if (index <= (HWEX_EXTENSIONS_COUNT >> 4)) { // check if the extension exists
+                if (index == (HWEX_EXTENSIONS_COUNT >> 4)) {
+                    // delete the flags of the non-existant extensions
+                    value = value & (0xFFFF >> (16 - (HWEX_EXTENSIONS_COUNT & 0xF)));
+                }
+                _GBAHardwareExtensionsIOWrite(gba, address, value);
+            }
+            break;
+        }
+        // Return codes
+        case REG_HWEX0_RET_CODE:
+            mLOG(GBA_IO, GAME_ERROR, "Write to read-only hardware extensions I/O register: %06X", address);
+		    break;
+
+        // CNT
+        case REG_HWEX0_CNT:
+            GBAHardwareExtensionsHandleCntWrite(gba, address, value);
+            break;
+
+        // Parameters
+        case REG_HWEX0_P0_LO:
+        case REG_HWEX0_P0_HI:
+        case REG_HWEX0_P1_LO:
+        case REG_HWEX0_P1_HI:
+        case REG_HWEX0_P2_LO:
+        case REG_HWEX0_P2_HI:
+        case REG_HWEX0_P3_LO:
+        case REG_HWEX0_P3_HI:
+            if (!(_GBAHardwareExtensionsIORead(gba, REG_HWEX0_CNT) & HWEX_CNT_FLAG_PROCESSING)) {
+                _GBAHardwareExtensionsIOWrite(gba, address, value);
+            }
+            break;
+        default:
+            mLOG(GBA_IO, GAME_ERROR, "Write non hardware extensions I/O register: %06X", address);
+    }
+}
+
+void GBAHardwareExtensionsIOWrite8(struct GBA* gba, uint32_t address, uint8_t value) {
+    uint32_t address16 = address & 0xFFFFFE;
+    uint16_t* reg = GetHwExIOPointer(gba, address16);
+    GBAHardwareExtensionsIOWrite(gba, address16, address & 1 ? (value << 8) | (*reg & 0xFF) : (value | (*reg & 0xFF00)));
+}
+
+void GBAHardwareExtensionsIOWrite32(struct GBA* gba, uint32_t address, uint32_t value) {
+    GBAHardwareExtensionsIOWrite(gba, address, value & 0xFFFF);
+    GBAHardwareExtensionsIOWrite(gba, address + 2, value >> 16);
+}
```

Comment:
If you don't have any special handling in these functions you should remove them; the regular IO handlers have support for that.

Reply:
I deleted the 32 bit one. But the 8bit one is needed or the default IO write function will try to write into the GBA structure directly.

Reply:
Read, not write, but I see what you mean. In that case this is fine.